index.bs

```diff
@@ -10699,14 +10714,18 @@ the <code>typeof</code> operator will return "function" when applied to an inter
         1.  Let &lt;|constructor|, |values|&gt; be the result
             of passing |S| and |args|.
             to the [=overload resolution algorithm=].
-        1.  Let |R| be the result of performing
-            the actions listed in the description of |constructor|
-            with |values| as the argument values.
+        1.  Let |targetRealm| be [$GetFunctionRealm$]({{NewTarget}}).
+        1.  Let |object| be the result of [=internally create a new object implementing the
```

Comment:
This looks wrong to me.

We want to create the object in the current Realm of the constructor, period.  That's how ES builtins work, last I checked, and we should not be inventing a new behavior.  All we should be using NewTarget for is the prototype object, which may end up coming from a different Realm.

In other words, we should be passing our current Realm, and then inside the "internally create" algorithm, if newTarget is not undefined, we can get the prototype from it, if an object use that, else fall back on the default proto in GetFunctionRealm(newTarget).  And fall back to the passed-in Realm if there is no newTarget.

Reply:
From what I can tell ES builtins actually create in the realm of NewTarget. See https://github.com/heycam/webidl/pull/635#discussion_r254879231 . I too was surprised.

Reply:
I used *realm* until you told me otherwise!

> Yeah, it looks like that's what ES does. Given that slightly-older Firefox releases and Safari do something different (use current Realm), it's not clear to me whether the spec changed there or what. But I agree we should just match ES.

Reply:
ES builtins get the default prototype to use if NewTarget's `.prototype` is not an object in the realm of NewTarget, which is what I thought https://github.com/heycam/webidl/pull/635#discussion_r254879231 was talking about.  But they create the object in the current Realm.  Which is not observable in ES so far (without weakrefs), because you can't observe a Realm for non-Function objects.  But if you look at the unrolled steps from ES, the ObjectCreate is happening at a point when the current Realm is the Realm of the constructor.

Note that ES is actually kinda internally inconsistent here.  In particular, if you look at https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-construct-argumentslist-newtarget (the [[Construct]] for normal scripted functions), it does the ObjectCreate before entering the Realm of the function itself.  So this creation happens in whatever Realm you're in when you call the constructor, which can be neither the Realm of the constructor nor the Realm of the NewTarget.  This came up recently in https://phabricator.services.mozilla.com/D19760#519322 and is likely a spec issue if object realms become observable.

On the other hand, for a built-in function (like Map, say, or any of the Web IDL constructors) the [[Construct]] is defined as https://tc39.github.io/ecma262/#sec-built-in-function-objects-construct-argumentslist-newtarget which just defers to https://tc39.github.io/ecma262/#sec-built-in-function-objects-call-thisargument-argumentslist with step 10 changed.  Importantly, there is no object-creation here outside of step 10 doing whatever it does, and we enter the Realm of the constructor in step 9.  So let's look at the Map constructor, to consider something specific.  It's defined at https://tc39.github.io/ecma262/#sec-map-iterable and calls https://tc39.github.io/ecma262/#sec-ordinarycreatefromconstructor which calls `ObjectCreate` while in the Realm of the Map constructor itself.  Again, the proto comes from whatever NewTarget is, but the actual object comes from the current Realm.

Reply:
Okay, thanks for clarifying. I wrote some tests, and it turns out no two browsers agree. I updated the PR to match my understanding of what ES does, based on your comment. Note that this doesn't match Gecko if NewTarget has a prototype property that isn't an object.

<table>
<thead>
<tr><th>Test <th>Gecko <th>WebKit <th>Chrome
<tbody>
<tr><th colspan=4>new DOMParser()
<tr><th>Prototype <td colspan=3>DOMParser.&shy;prototype
<tr><th>Object realm <td colspan=3>parent window
<tr><th colspan=4>new i.&shy;contentWindow.&shy;DOMParser()
<tr><th>Prototype <td colspan=3>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype
<tr><th>Object realm <td colspan=3>child window
<tr><th colspan=4>Reflect.&shy;construct(i.&shy;contentWindow.&shy;DOMParser, [], DOMParser)
<tr><th>Prototype <td>DOMParser.&shy;prototype <td>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype <td>DOMParser.&shy;prototype
<tr><th>Object realm <td>child window <td>child window <td>parent window
<tr><th colspan=4>Reflect.&shy;construct(DOMParser, [], i.&shy;contentWindow.&shy;DOMParser)
<tr><th>Prototype <td>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype <td>DOMParser.&shy;prototype <td>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype
<tr><th>Object realm <td>parent window <td>parent window <td>child window
<tr><th colspan=4>Reflect.&shy;construct(i.&shy;contentWindow.&shy;DOMParser, [], cons)
<tr><th>Prototype <td>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype <td>i.&shy;contentWindow.&shy;DOMParser.&shy;prototype <td>Object.&shy;prototype
<tr><th>Object realm <td>child window <td>child window <td>parent window
<tr><th colspan=4>Reflect.&shy;construct(DOMParser, [], i.&shy;contentWindow.&shy;cons)
<tr><th>Prototype <td>DOMParser.&shy;prototype <td>DOMParser.&shy;prototype <td>i.&shy;contentWindow.&shy;Object.&shy;prototype
<tr><th>Object realm <td>parent window <td>parent window <td>child window
<tr><th colspan=4>Conclusion
<tr><th>Prototype <td>From NewTarget; fallback to interface object from constructor's realm <td>Interface object from constructor's realm <td>From NewTarget; fallback to Object.prototype from NewTarget's realm
<tr><th>Object realm <td>From constructor <td>From constructor <td>From NewTarget
</table>

<details>
<summary>Test</summary>

```html
<!DOCTYPE html>
<script>
function identify(proto, i) {
  if (proto === Object.prototype)
    return "Object.prototype";

  if (proto === i.contentWindow.Object.prototype)
    return "i.contentWindow.Object.prototype";

  if (proto === DOMParser.prototype)
    return "DOMParser.prototype";

  if (proto === i.contentWindow.DOMParser.prototype)
    return "i.contentWindow.DOMParser.prototype";

  return "???"
}

function object_realm(dp) {
  var url = DOMParser.prototype.parseFromString.call(dp, "x", "text/html").documentURI;
  if (url === "http://software.hixie.ch/utilities/js/live-dom-viewer/")
    return "parent window";

  if (url === "http://software.hixie.ch/utilities/js/live-dom-viewer/document")
    return "child window";

  return "???";
}

function f(i) {
var dp = new DOMParser();
w("new DOMParser()");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

dp = new i.contentWindow.DOMParser();
w("new i.contentWindow.DOMParser()");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

dp = Reflect.construct(i.contentWindow.DOMParser, [], DOMParser);
w("Reflect.construct(i.contentWindow.DOMParser, [], DOMParser)");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

dp = Reflect.construct(DOMParser, [], i.contentWindow.DOMParser);
w("Reflect.construct(DOMParser, [], i.contentWindow.DOMParser)");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

var cons = function() {};
cons.prototype = 7;

dp = Reflect.construct(i.contentWindow.DOMParser, [], cons);
w("Reflect.construct(i.contentWindow.DOMParser, [], cons)");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

var script = i.contentWindow.document.createElement("script");
script.textContent = "window.cons = function() {}; cons.prototype = 8;";
i.contentWindow.document.documentElement.appendChild(script);

dp = Reflect.construct(DOMParser, [], i.contentWindow.cons);
w("Reflect.construct(DOMParser, [], i.contentWindow.cons)");
w(identify(dp.__proto__, i));
w(object_realm(dp));
w("");

}
</script>
<iframe src=document onload=f(this)></iframe>
```
</summary>